src/istream.cpp

```diff
@@ -0,0 +1,247 @@
+auto write_data_object(const po::variables_map& vm, io::client::default_transport& tp) -> int;
+
+int main(int argc, char* argv[])
+{
+    if (argc == 1) {
+        std::cerr << "Error: Invalid number of arguments.\n";
+        return 1;
+    }
+
+    po::options_description desc{"Options are"};
+    desc.add_options()
+        ("help,h", "Produce this message.")
+        ("offset,o", po::value<int_type>()->default_value(0), "The position of the read/write pointer.")
+        ("count,c", po::value<int_type>()->default_value(all_bytes), "The number of bytes to read/write.")
+        ("no-trunc", po::bool_switch(), "Do not truncate the data object on write.")
+        ("append,a", po::bool_switch(), "Append to the data object on write.")
+        ("stream_operation", po::value<std::string>(), "The stream operation (e.g. read or write).")
+        ("logical_path", po::value<std::string>(), "The logical path of a data object.");
+
+    po::positional_options_description pod;
+    pod.add("stream_operation", 1);
+    pod.add("logical_path", 1);
+
+    try {
+        po::variables_map vm;
+        po::store(po::command_line_parser(argc, argv).options(desc).positional(pod).run(), vm);
+        po::notify(vm);
+
+        if (const auto [return_ec, ec] = check_input_arguments(vm); return_ec) {
+            return ec;
+        }
+
+        auto& pack_table = irods::get_pack_table();
+        auto& api_table = irods::get_client_api_table();
+        init_api_table(api_table, pack_table);
+
+        rodsEnv env;
+        if (getRodsEnv(&env) < 0) {
+            std::cerr << "Error: Could not get iRODS environment.\n";
+            return 1;
+        }
+
+        irods::connection_pool conn_pool{1, env.rodsHost, env.rodsPort, env.rodsUserName, env.rodsZone, 600};
+        auto conn = conn_pool.get_connection();
+        io::client::default_transport tp{conn};
+
+        const auto stream_operation = vm["stream_operation"].as<std::string>();
+
+        if ("read" == stream_operation) {
+            return read_data_object(vm, tp);
+        }
+
+        if ("write" == stream_operation) {
+            return write_data_object(vm, tp);
+        }
+
+        std::cerr << "Error: Invalid stream operation [" << stream_operation << "]\n";
+
+        return 1;
+    }
+    catch (const std::exception& e) {
+        std::cerr << "Error: " << e.what() << '\n';
+        return 1;
+    }
+}
+
+auto usage() -> void
+{
+    std::cout << "Usage: istream read [-o INTEGER] [-c INTEGER] LOGICAL_PATH\n"
+                 "Usage: istream write [-o INTEGER] [-c INTEGER] [--no-trunc] [-a] LOGICAL_PATH\n"
+                 "\n"
+                 "Streams bytes to/from iRODS via stdin/stdout.\n"
+                 "Reads bytes from the target data object and prints them to stdout.\n"
+                 "Writes bytes from stdin to the target data object.\n"
+                 "\n"
+                 "If the target data object does not exist and --no-trunc is not passed,\n"
```

Comment:
i believe this is correct - just a very confusing triple negative...  let's play with the wording a bit.